### PR TITLE
ENH ufunc called on memmap return a ndarray

### LINKED ISSUE #667: regression (from 1.5.x) with return of operations on memmap'ed arrays (Trac #2215)
_Original ticket http://projects.scipy.org/numpy/ticket/2215 on 2012-09-14 by @yarikoptic, assigned to unknown._

As
http://projects.scipy.org/numpy/ticket/2179
stated: "Ideally one should be returning other ndarray objects, e.g. numpy scalars in the case of sum(), but this requires a larger rewrite. At least numpy version 1.5.1rc1 did this. The only time the _mmap should be kept (as I am aware of) is when slicing. "

This issue remains not fixed thus opening a separate issue for the above.  Taking into account that np.asscalar() called on scalars 
(e.g. which would be returned by np.sum on ndarrays) crashes, it becomes impossible to write code agnostic of ndarray/memmap container types.

it is also not clear why array-wide operations on mode='c' memmap'ed arrays which cannot really get reflected in the file storage, remain memmapped and do not result in normal ndarrays, but I guess that would be a separate issue...

### LINKED ISSUE #7403: np.average of memmap returns memmap 0d array instead of numpy scalar
```python
import numpy as np
mm = np.memmap('/tmp/test.data', mode='w+', shape=4)
np.average(mm)
```

Output since 5ceab8f982d80c68b3b5e90697cd20df51f4a3e8 (a couple of days ago):
```
memmap(0.0)
```

Output before that (e.g. in 1.10.4):
```
0.0
```

This causes the scikit-learn using numpy master tests to fail, see this Travis [log](https://travis-ci.org/scikit-learn/scikit-learn/jobs/114973679).

We have been bitten in the past by this in scikit-learn with `.sum` that has the same surprising behaviour (see https://github.com/scikit-learn/scikit-learn/pull/6225). Shouldn't reduce operations on a memmap array always return a numpy scalar ?

cc @ogrisel.

### PR DESCRIPTION
**Pull request by lesteve**
Special case for reduction functions (e.g. np.sum with axis=None) that
return a numpy scalar.

Fix #7403. Implemented `__array_wrap__` following @ahaldane's suggestion. Should fix #667 as well.

### CODE REVIEW

**Comment by seberg** [comment]
Years ago, had started something like this, but had to be a mit more complex, because of subclasses of memmap. You can find it here: https://github.com/numpy/numpy/pull/2932
Frankly, no idea why I closed it back then (did not check the posts), but it was a slow time for numpy IIRC, so might just have closed it because of that.

File: numpy/core/tests/test_memmap.py [hunk 1]
```diff
@@ -126,5 +128,23 @@ def test_view(self):
...
+        for unary_op in [sum, average, product]:
+            result = unary_op(fp)
+            assert_(isscalar(result))
+            assert_(result.__class__ is self.data[0, 0].__class__)
+
+            assert_(unary_op(fp, axis=0).__class__ is ndarray)
+            assert_(unary_op(fp, axis=1).__class__ is ndarray)
+
+        for binary_op in [add, multiply, divide, add]:
+            assert_(binary_op(fp, self.data).__class__ is ndarray)
```
Comment (charris) [inline numpy/core/tests/test_memmap.py]: I suspect that the test failure is because `self.data` starts with zero, which causes problems with division.

**Comment by charris** [comment]
Not sure if the counts as a bug/enhancement/change in behavior. If the latter, might need to start with a `FutureWarning`.

**Comment by ahaldane** [comment]
Looks good, though I'm pretty sure in-place operations (eg `m += 1`) don't work properly, and  I didn't immediately see a good fix for that. I'd also need to think a little more to make sure there are no other broken edge cases.

Also, I wonder if this is an opportunity to try out the new `__numpy_ufunc__` instead of `__array_wrap__`. As I understand `__numpy_ufunc__` is meant to replace it, although it's not totally done yet. Just another thing to ponder.

**Comment by ahaldane** [comment]
Well, looks like `__numpy_ufunc__` was disabled for now, so we shouldn't try that.

Here's a possible fix for `__array_wrap__` to make in-place operations retain the subtype:
```    
def __array_wrap__(self, arr, context=None):
        if np.may_share_memory(self, arr):
            return arr
        if arr.shape == ():
            return arr[()]
        return arr.view(ndarray)
```

so now you can also do
```
>>> a = memmap(np.memmap('file', dtype='i4', shape=(4,4))
>>> a += 1
>>> a.flush()
```

More generally, there are 2 things I am not 100% sure of yet which we need to be more sure of:

1. Are we sure that there are no ufunc uses which *should* return a 0-d array instead of a scalar?
2. Are we sure that there are no cases where `np.may_share_memory(self, arr) == False` but we should actually be returning a memmap? That is, where arr's memory is actually a memmapped buffer?

**Comment by ahaldane** [comment]
Oh, I just read through @seberg's PR, which looks like it addresses more of these concerns. @lesteve @seberg maybe we should revive that one? Even though it does seem a little dangerous, I think it might be worth doing since we have a couple people now who would like the change.

**Comment by seberg** [comment]
Dunno. Don't even know what my code was compared to this (why did I
have array_prepare also?). I am relatively sure you don't really have
to worry much about the mmapped buffers or so. If `arr` does not share
memory with self, then the array should be freshly created with its own
data. If an "out" array exists, the "out" array gets to do the
array_wrap (maybe that is why array_prepare, hmmm).

Only thing I would be worried about that memmaps have some attributes
that still somewhat work when we now would return arrays, and thus
might break code, that just use them for no good reason. Should check,
but I could imagine it is fine, since a memmap object that (I guess?)
errors on flush is pretty useless.

About 1. Yes, we are sure ufuncs never return 0-d arrays (except for
subclasses to give subclasses a chance and even that changed probably
in 1.7 or so. That does not mean I agree with it, but Numpy doesn't
consider 0-d arrays first class citizens.

Anyway, if we are not sure about all of these things, possibly we could
just change the 0-d case, and hope that since the behaviour was changed
(maybe not so long ago?) it is fine.

**Comment by njsmith** [comment]
Don't have time to read through all of this in detail right now, but as a general note I'm +10 on making it so that memmaps don't propagate through ufuncs. I'd also vote against requiring a `FutureWarning`, because (a) this will be really noisy and annoying, because it would trigger every time you do anything with memmaps, even totally safe things, and there's no reasonable way to change existing code to avoid the warning, (b) the only behavioural difference between an `np.memmap` object and an `np.ndarray` is that the former has a `._mmap` attribute that can be `flush`ed, but even this is not true for the `np.memmap` objects that get allocated and returned from ufuncs -- they are literally just `ndarray` objects with a confusing subclass attached.

**Comment by charris** [comment]
Should get a mention in the 1.12.0-notes.

**Comment by charris** [comment]
@lesteve Ping.

**Comment by lesteve** [comment]
I added something in doc/release/1.12.0-notes.rst.

I haven't had time to work on this in the past few days. I will try to understand better what was done in #2932.

**Comment by ahaldane** [comment]
@lesteve My impression of the difference between this PR and #2932 is that #2932 tries to treat subclasses of memmap specially and handles `__getitem__` (but it missing the 0d array case in this PR), while this PR treats 0d arrays properly but doesn't worry about inheritance or `__getitem__`.

But I will admit I don't really understand a lot of  #2932. It appears to be trying to take care of multiple inheritance, involving inheritance from `memmap`. But I'm not sure it's worth the effort to do that. Do we know of any projects that use multiple inheritance involving `memmap`? Note also that this PR, as is, properly takes care of single inheritance of `memmap` (since subclasses will inherit `array_wrap`).

However, I do think #2932 has the right hint in `__getitem__`: I think we want fancy indexing like
```
>>> a = memmap(np.memmap('file', dtype='i4', shape=(4,4))
>>> a[[0,3]]
```
to return an ndarray, since the result is a new array rather than a view. However `a[1:3]` should give back a memmap . I am not sure how to implement that yet.

In summary, my impression is that we can forget about trying to fix multiple inheritance like #2932, however we should somehow take care of fancy indexing, and I also suggest adding the `if np.may_share_memory(self, arr):` clause I pasted above.

**Comment by lesteve** [comment]
I have added the `__getitem__` behaviour + tests from #2932 and also the `may_share_memory` in `__array_wrap__`. I think it does what @ahaldane wanted for view vs new array for `__getitem__`.

I haven't tried to understand further the subtleties involved with subclassing that #2932 was trying to tackle.

**Comment by ahaldane** [comment]
I think this PR is ready to be merged as is, if there are no objections. 

Although I have a little bit of a feeling there may be some "unknown unknowns" here, after sitting on it I haven't though up any more cases to take care of, and I'm not too worried about the multiple subclass case. Plus, I think this is easy to undo later if there are problems, and it can get some testing if we merge.

Maybe we can wait a little for anyone else to speak up, and then merge.

File: numpy/core/memmap.py [hunk 2]
```diff
@@ -309,3 +309,21 @@ def flush(self):
         """
         if self.base is not None and hasattr(self.base, 'flush'):
             self.base.flush()
+
+    def __array_wrap__(self, arr, context=None):
+        # Return scalar instead of 0d memmap, e.g. for np.sum with
+        # axis=None
+        if arr.shape == ():
```
Comment (seberg) [inline numpy/core/memmap.py]: Just to be on the safe side, it might be good to move it after the may_share_memory call? I don't think we should ever return 0-d arrays from functions calling array_wrap, but I am not perfectly sure.

Comment (seberg) [inline numpy/core/memmap.py]: Then again, maybe it doesn't matter.

**Comment by seberg** [comment]
If there may be problems with subclasses, I would feel better if there is a note to the mailing list.

**Comment by seberg** [comment]
Otherwise, I trust your assessment :).

File: numpy/core/memmap.py (hunk 2 above)
Comment (lesteve) [inline numpy/core/memmap.py]: > Just to be on the safe side, it might be good to move it after the may_share_memory call? 

Done.

**Comment by lesteve** [comment]
I am on holiday for a week with a very spotty internet connection. I'll 
do that when I get back home.

**Comment by lesteve** [comment]
> I'll do that when I get back home.

OK I rebased on master and fixed the conflicts. Is there anything else I can do ?

**Comment by ahaldane** [comment]
All right, let me send an email to the list (unless you would like to) to try and fish out any problems involving multiple inheritance, and if there are no comments in a day or two let's merge.

**Comment by lesteve** [comment]
Go for it, you'll be more able to explain the problems that are likely to arise.

**Comment by mhvk** [comment]
I hate to be a spoilsport so late in a commit, but I think you're missing a possibly important case here, viz., that someone has set `out` in the call to the `ufunc`; in this case, you should definitely not `view` it as something else! Note that you avoid the view already in the most likely case, of in-place operations (`mm += 1` is translated to `np.add(mm, 1, out=mm)`, but not a case where someone has set up a different memmap output array (say, `np.add(mm1, 1, out=mm2)`). Furthermore, suppose `out` is a different subclass of `ndarray`, perhaps a masked array or so, then the user will be surprised that `np.add(mm1, 1, out=ma)` will now return a normal `ndarray` (even though `ma` will still be a masked array). 

I have not thought things through perhaps well enough, but it may be this can be resolved by introducing an `__array_prepare__` instead of an `__array_wrap__` -- at least this is where we change ndarray output to `Quantity` in astropy... In any case, note that to look for in-place, you can simply check `arr is self`, no need for `np.may_share_memory`, though it may be best to change to a similar logic to that in `__getitem__`.

**Comment by lesteve** [comment]
> I hate to be a spoilsport so late in a commit, but I think you're missing a possibly important case here

No problem and thanks a lot for your feedback !

https://github.com/numpy/numpy/pull/2932 was implementing `__array_prepare__` maybe this is for this kind of reasons. I'll investigate in more details.

**Comment by lesteve** [comment]
@mhvk I don't seem to see the problem you describe unless I am missing something.

Here is a snippet:
```python
import numpy as np

rng = np.random.RandomState(0)

mm0 = np.memmap('mm0.dat', mode='w+', dtype='f8', shape=(3, 4))
mm0[:] = rng.randn(3, 4)
mm1 = np.memmap('mm1.dat', mode='w+', dtype='f8', shape=(3, 4))

np.add(mm0, 1, out=mm1)
print(repr(mm1))
```

If I understood correctly you thought mm1 would be somehow turned from a memmap into an array. This is not what I am seeing, the output of the snippet is:

```
memmap([[ 2.76405235,  1.40015721,  1.97873798,  3.2408932 ],
       [ 2.86755799,  0.02272212,  1.95008842,  0.84864279],
       [ 0.89678115,  1.4105985 ,  1.14404357,  2.45427351]])
```

**Comment by embray** [comment]
I haven't reviewed the change closely but it's a much welcome change in principle so :+1: (in the long term I think the `memmap` class should go away entirely--some of its code is worth saving, but I don't think it makes much sense as a separate ndarray subclass).

**Comment by mhvk** [comment]
@lesteve - the problem I mentioned is that in your example, `np.add` *returns* an ndarray. Normally, when `out` is given, it is ensured that what is returned is `out` itself, i.e., if one does `mm2 = np.add(mm0, 1, out=mm1)`, it should be the case that `mm2 is mm1`.

**Comment by lesteve** [comment]
> `mm2 = np.add(mm0, 1, out=mm1)`, it should be the case that `mm2 is mm1`

@mhvk this PR does have the behaviour you want. Thanks for the explanations, I naively expected `np.add` to return None when the `out` parameter was specified.

**Comment by mhvk** [comment]
OK, I now see why this does work (should have seen it before): when the output is an memmap, it is `out.__array_wrap__` that is called, so inside `__array_wrap__`, the condition `may_share_memory` is in fact true.

File: numpy/core/memmap.py [hunk 3]
```diff
@@ -309,3 +309,20 @@ def flush(self):
         """
         if self.base is not None and hasattr(self.base, 'flush'):
             self.base.flush()
+
+    def __array_wrap__(self, arr, context=None):
+        # Return memmap if self and arr are likely to share memory
+        if np.may_share_memory(self, arr):
```
Comment (mhvk) [inline numpy/core/memmap.py]: So, a remaining suggestion is to replace this statement with the simpler (and faster)
```
# Return the object it is was given as an output and should thus preserve type
if arr is self:
    return arr
```

File: numpy/core/memmap.py [hunk 4]
```diff
@@ -309,3 +309,20 @@ def flush(self):
         """
         if self.base is not None and hasattr(self.base, 'flush'):
             self.base.flush()
+
+    def __array_wrap__(self, arr, context=None):
+        # Return memmap if self and arr are likely to share memory
```
Comment (mhvk) [inline numpy/core/memmap.py]: p.s. Looking at #2932, I think it would be nice to properly account for possibly being not directly an `ndarray` subclass (but, e.g., a mixed type like `class(Quantity, memmap)`, i.e., I think I would here include the following
```
    arr = super(memmap, self).__array_wrap__(arr, *args, **kwargs)
    # Return arr if it not a memmap or if it was given as output.
    if arr is self or type(arr) is not memmap:
        return arr
    # If we were created in the `ufunc`, we should not be a memmap,
    # but return a scalar or ndarray, as appropriate.
    <your lines>
```

File: numpy/core/memmap.py (hunk 3 above)
Comment (lesteve) [inline numpy/core/memmap.py]: Playing a bit around, I wasn't able to find an edge case in `__array_wrap__` where `np.may_share_memory(self, arr)` and `arr is self` disagreed, so I made the advised change.

If someone else can see an edge case that I missed, please complain !

**Comment by ahaldane** [comment]
Thanks for the sugestions  @mhkv, it looks better now.

@lesteve I want to consider @mhkv's other suggestion too, to use super, but I am not totally happy with that solution because it breaks the single inheritance case (eg `class B(np.memmap)`). I want to think about it a little more.

**Comment by lesteve** [comment]
Let me know, I won't be able to work on it until Monday though.

**Comment by ahaldane** [comment]
All right, from a brief look, astropy code doesn't seem to subclass memmap anywhere so @mhvk's example of `class MyClass(Quantity, memmap)` is only hypothetical.

In that case I vote to keep things simple and ignore the subclass issue, and merge this as is. Is that all right @mhvk?

Anyone in the future who wants to subclass memmap will be able to override our behavior if they want.

**Comment by mhvk** [comment]
@ahaldane - I'm half curious about what goes wrong with the inheritance `class B(np.memmap)`, but as my example was indeed hypothetical, I'm quite happy to keep it simple.

**Comment by ahaldane** [comment]
@mhvk, maybe I am mistaken. I was bothered by the fact that ufuncs and fancy slices of instances of `B` wouldn't return ndarrays, eg. `np.mean(B(file, dtype='i4', shape=4))` returns a `B` scalar. But actually that is probably what should happen for subclasses of ndarray.

@lesteve, sorry for going back and forth on this, but could you make the `super` change?
```
    def __array_wrap__(self, arr, context=None):
        arr = super(memmap, self).__array_wrap__(arr, context)
        # Return arr if it not a memmap or if it was given as output.
        if arr is self or type(arr) is not memmap:
            return arr
        ...
```

and a unit test could be
```
class B(np.memmap):
    pass

a = B('file', dtype='i4', shape=4)
assert_equal(type(np.mean(a)), B)
assert_equal(type(a[[0,1]]), B)
```

**Comment by lesteve** [comment]
> I was bothered by the fact that ufuncs and fancy slices of instances of B wouldn't return ndarrays, eg. np.mean(B(file, dtype='i4', shape=4)) returns a B scalar. But actually that is probably what should happen for subclasses of ndarray.

I find that slightly bothering too. In my mind, it is very unlikely that one would want a B 0d array for the result of np.mean and for fancy slices I am a bit more split but not really convinced either. I'd be interested if someone has some kind of use case where this behaviour makes sense.

I made the requested change anyway.

**Comment by ogrisel** [comment]
> I haven't reviewed the change closely but it's a much welcome change in principle so :+1: (in the long term I think the memmap class should go away entirely--some of its code is worth saving, but I don't think it makes much sense as a separate ndarray subclass).

@embray I agree. To fix the `memmap` class consistency issue I had prototyped in the past an alternative that would not subclass ndarray but instead call `np.frombuffer` on a custom mmap buffer that would store the required metadata (filename, offset, mode...):

https://github.com/joblib/joblib/commit/b68a86891f6a90dfe74043e59cfb9b5f50b738de#diff-19004d02cec64e3d448b33565de338ebR94

I never finished that work but if people are interested I could contribute it to numpy as a saner alternative to `np.memmap`. Do you want me to extract that old code and make it a PR for numpy (with tests)?

**Comment by embray** [comment]
@ogrisel Yes, that looks useful.  For myself, I'm using `np.memmap` to create the underlying mmap in a convenient way, but then immediately passing it to `np.frombuffer` and throwing away the np.memmap` object.

**Comment by mhvk** [comment]
@lesteve - for the subclass case, I do think it is best to default to returning the subclass. It is quite possible that the subclass stores information that is useful even for a 0-d array (e.g., a `Quantity` stores a unit, which should not get lost).

Overall, it seems to me this is a good addition as is, and with all the tests, etc., it seems ready to go in.

**Comment by lesteve** [comment]
> @lesteve - for the subclass case, I do think it is best to default to returning the subclass. It is quite possible that the subclass stores information that is useful even for a 0-d array (e.g., a Quantity stores a unit, which should not get lost).

Fair enough, thanks for the explanation!

**Comment by ahaldane** [comment]
Looks good, I'll merge in an hour or two.

**Comment by ahaldane** [comment]
Merged, thanks @lesteve and the many reviewers.

**Comment by lesteve** [comment]
Great, thanks a lot !
